Main.py

```diff
@@ -33,306 +35,363 @@
+
+    world.difficulty_requirements = difficulties[world.difficulty]
+
+    for player in range(1, world.players + 1):
+        create_regions(world, player)
+        create_dungeons(world, player)
+
+    logger.info('Shuffling the World about.')
+
+    for player in range(1, world.players + 1):
+        link_entrances(world, player)
+
+    mark_light_world_regions(world)
+
+    logger.info('Generating Item Pool.')
+
+    for player in range(1, world.players + 1):
+        generate_itempool(world, player)
+
+    logger.info('Calculating Access Rules.')
+
+    for player in range(1, world.players + 1):
+        set_rules(world, player)
+
+    logger.info('Placing Dungeon Prizes.')
+
+    fill_prizes(world)
+
+    logger.info('Placing Dungeon Items.')
+
+    shuffled_locations = None
+    if args.algorithm in ['balanced', 'vt26'] or args.keysanity:
+        shuffled_locations = world.get_unfilled_locations()
+        random.shuffle(shuffled_locations)
+        fill_dungeons_restrictive(world, shuffled_locations)
+    else:
+        fill_dungeons(world)
+
+    logger.info('Fill the world.')
+
+    if args.algorithm == 'flood':
+        flood_items(world)  # different algo, biased towards early game progress items
+    elif args.algorithm == 'vt21':
+        distribute_items_cutoff(world, 1)
+    elif args.algorithm == 'vt22':
+        distribute_items_cutoff(world, 0.66)
+    elif args.algorithm == 'freshness':
+        distribute_items_staleness(world)
+    elif args.algorithm == 'vt25':
+        distribute_items_restrictive(world, 0)
+    elif args.algorithm == 'vt26':
+
+        distribute_items_restrictive(world, gt_filler(world), shuffled_locations)
+    elif args.algorithm == 'balanced':
+        distribute_items_restrictive(world, gt_filler(world))
+
+    if world.players > 1:
+        logger.info('Balancing multiworld progression.')
+        balance_multiworld_progression(world)
+
+    logger.info('Patching ROM.')
+
+    if args.sprite is not None:
+        if isinstance(args.sprite, Sprite):
+            sprite = args.sprite
+        else:
+            sprite = Sprite(args.sprite)
+    else:
+        sprite = None
+
+    player_names = parse_names_string(args.names)
+    outfilebase = 'ER_%s_%s-%s-%s%s_%s-%s%s%s%s%s_%s' % (world.logic, world.difficulty, world.mode, world.goal, "" if world.timer in ['none', 'display'] else "-" + world.timer, world.shuffle, world.algorithm, "-keysanity" if world.keysanity else "", "-retro" if world.retro else "", "-prog_" + world.progressive if world.progressive in ['off', 'random'] else "", "-nohints" if not world.hints else "", world.seed)
+
+    use_enemizer = args.enemizercli and (args.shufflebosses != 'none' or args.shuffleenemies or args.enemy_health != 'default' or args.enemy_health != 'default' or args.enemy_damage or args.shufflepalette or args.shufflepots)
+
+    jsonout = {}
+    if not args.suppress_rom:
+        from MultiServer import MultiWorld

```

Comment:
Not sure if we need something along the lines of multi-data, but I'll want to rip it out at first.